operators/src/pro/cache/cache_chunks.rs

```diff
@@ -0,0 +1,398 @@
+impl_cache_element_subtype_magic!(MultiPoint, MultiPoint);
+impl_cache_element_subtype_magic!(MultiLineString, MultiLineString);
+impl_cache_element_subtype_magic!(MultiPolygon, MultiPolygon);
+
+/// Our own tile stream that "owns" the data (more precisely a reference to the data)
+#[pin_project(project = CacheChunkStreamProjection)]
+pub struct CacheChunkStream<G> {
+    data: Arc<Vec<FeatureCollection<G>>>,
+    query: VectorQueryRectangle,
+    idx: usize,
+}
+
+impl<G> CacheChunkStream<G>
+where
+    G: Geometry,
+    FeatureCollection<G>: FeatureCollectionInfos,
+{
+    pub fn new(data: Arc<Vec<FeatureCollection<G>>>, query: VectorQueryRectangle) -> Self {
+        Self {
+            data,
+            query,
+            idx: 0,
+        }
+    }
+
+    pub fn element_count(&self) -> usize {
+        self.data.len()
+    }
+}
+
+impl<G: Geometry> Stream for CacheChunkStream<G>
+where
+    FeatureCollection<G>: CacheElementHitCheck,
+    G: ArrowTyped,
+{
+    type Item = Result<FeatureCollection<G>>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let CacheChunkStreamProjection { data, query, idx } = self.as_mut().project();
+
+        // return the next tile that is contained in the query, skip all tiles that are not contained
+        for i in *idx..data.len() {
+            let chunk = &data[i];
+            if chunk.cache_element_hit(query) {
+                // TODO: we really should cache the elements bbox somewhere
+                let Ok(chunk) = chunk.filter_cache_element_entries(query) else {
+                    // This should not happen, since we already checked that the element is contained in the query
+                    log::error!("Could not filter cache element entries");
+                    continue;
+                };
+
+                // if the chunk is empty, we can skip it
+                if chunk.is_empty() {
+                    log::trace!("Skipping empty chunk after filtering for query rectangle");
+                    continue;
+                }
+
+                // set the index to the next element
+                *idx = i + 1;
+                return std::task::Poll::Ready(Some(Ok(chunk)));
+            }
+        }
+
+        std::task::Poll::Ready(None)
+    }
+}
+
+pub enum TypedCacheChunkStream {
+    NoGeometry(CacheChunkStream<NoGeometry>),
+    MultiPoint(CacheChunkStream<MultiPoint>),
+    MultiLineString(CacheChunkStream<MultiLineString>),
+    MultiPolygon(CacheChunkStream<MultiPolygon>),
+}
+
+pub trait CacheElementHitCheck {
+    fn cache_element_hit(&self, query_rect: &VectorQueryRectangle) -> bool;
+
+    fn filter_cache_element_entries(
+        &self,
+        query_rect: &VectorQueryRectangle,
+    ) -> Result<Self, CacheError>
+    where
+        Self: Sized;
+}
+
+impl CacheElementHitCheck for FeatureCollection<NoGeometry> {
```

Comment:
what is this trait used for, can you add a comment?

Why does it only use the time bounds and not the bbox?